src/tf/parse_batchnorm.cpp

```diff
@@ -0,0 +1,34 @@
+#include <migraphx/tf/op_parser.hpp>
+#include <migraphx/tf/tf_parser.hpp>
+#include <migraphx/ranges.hpp>
+#include <migraphx/make_op.hpp>
+
+namespace migraphx {
+inline namespace MIGRAPHX_INLINE_NS {
+namespace tf {
+
+struct parse_batchnorm : op_parser<parse_batchnorm>
+{
+    std::vector<op_desc> operators() const { return {{"FusedBatchNorm"}, {"FusedBatchNormV3"}}; }
+
+    instruction_ref parse(const op_desc& /*opd*/,
+                          const tf_parser& parser,
+                          tf_parser::node_info info,
+                          std::vector<instruction_ref> args) const
+    {
+        args = parser.to_nchw(args);
```

Comment:
I really would like to avoid having to explicitly transposing the inputs and outputs. There should be a field such as `transpose` that can be set to enable transposing implicitly. We can do this by defining a `base_parse` function in the `op_parser` base class:

```cpp
template<class Derived>
struct op_parser : auto_register<register_op_parser_action, Derived>
{
    bool transpose = false;

    std::vector<instruction_ref> base_parse(const op_desc& opd,
                          const tf_parser& parser,
                          tf_parser::node_info info,
                          const std::vector<instruction_ref>& args) const
    {
        std::vector<instruction_ref> result;
        auto& self = static_cast<const Derived&>(*this);
        if (self.transpose)
        {
            result = implicit_multi_op(self.parse(opd, parser, info, parser.to_nchw(args)));
            std::transform(result.begin(), result.end(), result.begin(), [&](auto ins) {
                return parser.to_nhwc(ins);
            });
        }
        else
        {
            result = implicit_multi_op(self.parse(opd, parser, info, args));
        }
        return result;
    }
};
```

So classes can just do: 

```cpp
struct parse_batchnorm : op_parser<parse_batchnorm>
{
    bool transpose = true;
    std::vector<op_desc> operators() const { return {{"FusedBatchNorm"}, {"FusedBatchNormV3"}}; }

    instruction_ref parse(const op_desc& /*opd*/,
                          const tf_parser& parser,
                          tf_parser::node_info info,
                          std::vector<instruction_ref> args) const;
};
```

The register function would be changed to call `base_parse`:

```cpp
template <class T>
void register_op_parser()
{
    T parser;
    for(auto&& opd : parser.operators())
        register_op_parser(opd.tf_name, [opd, parser](auto&&... xs) {
            return parser.base_parse(opd, xs...);
        });
}
```

Reply:
changed